### PR TITLE
 [FLINK-26516][streaming] Recover GlobalCommittables with Sink V1 GlobalCommittable serializer

### PR DESCRIPTION
**Pull request by fapaul**
## What is the purpose of the change

    With SinkV2 the committer and global committer work very similar and
    they only write committables into state. SinkV1's GlobalCommitter on the
    other hand used to write GlobalCommittables into state so this commits
    adds an migration path.

## Brief change log

- c725569 enabled the GlobalCommitter IT cases unfortunately we missed to enable them after FLINK-25726 was merged
- 8549f8f Pass the global committer and serializer to the global committer operator to restore Sink V1 state

## Verifying this change

- Reenabled tests pass
- Added an upgrade test to restore a savepoint from Sink V1.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by fapaul** [comment]
@gaoyunhaii I think this should fix the upgrade path from Sink V1 to Sink V2.

**Comment by fapaul** [comment]
@flinkbot run azure

**Comment by fapaul** [comment]
@flinkbot run azure

**Comment by gaoyunhaii** [comment]
Thanks @fapaul for the PR! I'll have a look~

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java
Comment (gaoyunhaii) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:146]: Might move this statement into `context.isRestored`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java
Comment (gaoyunhaii) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:227]: Might add `checkState(checkpointId > InitContext.INITIAL_CHECKPOINT_ID)`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java
Comment (gaoyunhaii) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java:451]: Here seems should be `failures.forEach`?

**Comment by gaoyunhaii** [comment]
Very thanks @fapaul for the PR!

One main issue from my side is that it seems the behavior of the current `GlobalCommitterOperator` is still not compatible with the legacy one? Currently once the legacy state is restored, in the following execution the behavior of `combine`, `filterRecoveredCommittables` and `endOfInput` is still not called~?

**Comment by fapaul** [comment]
> Very thanks @fapaul for the PR!
> 
> One main issue from my side is that it seems the behavior of the current `GlobalCommitterOperator` is still not compatible with the legacy one? Currently once the legacy state is restored, in the following execution the behavior of `combine`, `filterRecoveredCommittables` and `endOfInput` is still not called~?

I would have though `combine` and `filterRecoveredCommittables` are now equal or very closely to what was done before.

To be honest I never really understood the idea behind exposing `endOfInput` of the GlobalCommitter. I also have never seen a sink really using it. Can you maybe provide more context here? I would probably make a big warning in the release notes that this `endOfInput` is not used anymore.

**Comment by gaoyunhaii** [comment]
Hi Fabian~

> I would have though combine and `filterRecoveredCommittables` are now equal or very closely to what was done before.

It seems currently `filterRecoveredCommittables` and `combine` are used only on the legacy state~? For example, suppose we recovered from sink v1, then on restoring we would call `filterRecoveredCommittables`, then if the job continue running and take another several checkpoints, then after failover all committables are stored in the `CommittableCollector` and on restoring `filterRecoveredCommittables` seems would not be called~?

For `endOfInput` it is initially designed to write metas, like write partition to the hive metastore or write `_SUCCESS` file under the directory. I'm also agree with we could change it if no users are using it since it is marked as `Experimental` before.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java
Comment (fapaul) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:227]: I guess it could be that the pipeline is started from a savepoint that is `InitContext.INITIAL_CHECKPOINT_ID`. I think checking `checkpointId >= InitContext.INITIAL_CHECKPOINT_ID` makes more sense. WDYT?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java
Comment (fapaul) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java:451]: This is not easily possible because only committable are retriable at this point. I would need to add a special logic when this happens. I rely here on the fact that I assume all `commit` calls are idempotent.

**Comment by fapaul** [comment]
> It seems currently `filterRecoveredCommittables` and `combine` are used only on the legacy state~? For example, suppose we recovered from sink v1, then on restoring we would call `filterRecoveredCommittables`, then if the job continue running and take another several checkpoints, then after failover all committables are stored in the `CommittableCollector` and on restoring `filterRecoveredCommittables` seems would not be called~?

I thought about this for a longer time and think the behaviour of `combine` should be fine since it is always used when calling commit. The semantic of `filterRecoveredCommittables` is definitely currently unclear although I would also question the purpose of this method.
In general, I think all commit calls be for the committer of global committer need to be idempotent and handling already committed committables needs to happen in the commit method either way. I do not see an immediate benefit of calling `filterRecoveredCommittables`.

I am not really what to do about that. Do you have an idea?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java:451]: Thanks for the clarification! Might we add some comments for this~?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java
Comment (gaoyunhaii) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java:227]: Since it is called from `prepareSnapshotPreBarrier`, thus if the job is restored from savepoint, the next checkpoint should start from at least the `restored checkpoint id + 1`, thus it should always larger than `InitContext.INITIAL_CHECKPOINT_ID` ?

The main issues of `checkpointId == InitContext.INITIAL_CHECKPOINT_ID` is that in this case there will be two summary messages of the same checkpoint id, which would cause exceptions in `CommitterOperator`.

**Comment by gaoyunhaii** [comment]
Hi Fabian~ thanks for the clarification!

For the `filterRecoveredCommittables`, it should be initiated by some data lake systems like iceberg, if there are global committables from more than one checkpoint not committed before failover, then it should be able to merge them into one global committable on restoring, thus the restoring process could be accelerated [1].

For the `combine`, sorry I overlooked the issue and the current implementation, currently it is indeed called before committed. The issue is said that the committables should be combined on `snapshotState`, and the resulted global committables should be persisted, thus if there are failovers and re-commit, the `combine` does need to be executed again.

If we want to call `filterRecoveredCommittables` for each snapshot, we might need to add a parameter `isRestored` to the `CommittableManager#commit`. If we want to also keep the same behavior for the `combine` method, we might need to do some more modification to the `CommittableManager` or not reuse the current implementation.

However, currently the sink implementation in the `iceberg`, `hudi` and `flink-table-store` are indeed using the customized operators / legacy sink API instead of the sink API v1. In addition, the two issues should mainly affect he performance instead of the correctness. At last, users are still be able to add their own global committer implementation with topology API. Thus I think it is still useful to solve the two issues, but if we do not have capacity we might postpone solving the two issues. 

If so, perhaps we could create a new jira issue for state compatibility with the sink v1 and attach this PR to that issue, and dowgrade the FLINK-26173 and leaves it open? Since we have not fully fix the issues listed in the FLINK-26173~ 

[1] https://lists.apache.org/thread/3tqgnhclbr8ptb69b2ro74535dtbd608

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java
Comment (fapaul) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/SinkV1Adapter.java:451]: Good idea, I add a comment.

**Comment by fapaul** [comment]
> If so, perhaps we could create a new jira issue for state compatibility with the sink v1 and attach this PR to that issue, and dowgrade the [FLINK-26173](https://issues.apache.org/jira/browse/FLINK-26173) and leaves it open? Since we have not fully fix the issues listed in the [FLINK-26173](https://issues.apache.org/jira/browse/FLINK-26173)~

Thanks for the suggestion. I created a new ticket https://issues.apache.org/jira/browse/FLINK-26516 to fix the state incompatibility and lowered the priority of the original ticket.

**Comment by gaoyunhaii** [comment]
I added a fix for the failed azure test: previously if there are no committables the combine and commit would be skipped, otherwise there might create an empty global committable. The test failed due to this reason.

**Review by gaoyunhaii — APPROVED** [review]
Thanks @fapaul for the PR! LGTM

**Pull request closed without merging**
